Start team-based exercises and create submissions

madwau opened the pull request:
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
This PR is part of the epic of bringing team-based exercises to Artemis. So far, instructors can create team-based exercises and manage teams per exercise.

With this PR, students can start team-based exercises as they would start a regular exercise, only that there will be just one shared participation for all the students of a team. Thus, only one of the students of a team will actually start the exercise while the others use the then existing participation. Any one of the team members can then contribute in the form of submissions.

For programming exercises, all team members will have access to the repository. When the members of a team are changed by a TA or instructor, the permissions will be updated accordingly. Removed team members will be switched to read-only access (this is the case for Bitbucket and needs to be improved soon so that they lose access completely; for Gitlab they are already fully removed). New team members are given access to the repository.

### Steps for Testing
1. Log in to Artemis
2. Navigate to Course Administration
3. Click on a Course
4. Create a new Text, Modeling or File Upload Exercise and pick the mode `TEAM` (only accessible to `ROLE_ADMIN`)
5. Go to the Exercise overview for the course
6. Click on the button "Teams" for your newly created team exercise
7. Create a few teams (also play around with updating and deleting teams)
8. Log in as one of the students who is part of a team
9. Check that the dashboard shows the team exercise and that the buttons "Team" and "Start exercise" are available
10. Click on the button "Team" to check out who your team members are
11. Go back and start the exercise
12. Create a submission
13. Log in as admin again and look at the participation and submission
14. Create a result for the submission
15. Log in as the team member again and check that the result is available

### Steps for Testing: Programming exercises
1. Log in to Artemis
2. Navigate to Course Administration
3. Click on a Course
4. Create a new Programming Exercise and pick the mode `TEAM` (only accessible to `ROLE_ADMIN`)
5. Go to the Exercise overview for the course
6. Click on the button "Teams" for your newly created team exercise
7. Create a few teams (also play around with updating and deleting teams)
8. Log in as one of the students who is part of a team
9. Submit code via the Online Editor and via cloning the repository locally
10. Check that everything works as usual

### Description
Note: The numbering here corresponds to the numbering in the screenshot section.

1. Improvement to Team Management page for exercise
    * Team's short name is linked to the new Team Detail page
    * The students are linked to their user management page
    * Better looking horizontal scrollbar for the students list (custom scrollbar only for webkit browsers)
    * Clicking on "Delete" now opens a confirmation modal
2. Improved search on Team Management page
    * Search is now possible for team (short) name and students's login and name
    * Autocomplete not only shows team name but also the students
3. ShortName was added to the Team model
    * It is used similarly to how login is used for students (acts as a unique identifier, e.g. in repository urls or build plans)
    * It cannot be changed after a team was created
4. Delete confirmation modal was added for deleting teams
    * Wrapped in TeamDeleteButtonComponent (which is used on the team overview and detail page)
5. Team Detail page was added
    * Simply shows a list of students that are part of the team
    * This page can be accessed by students for the teams that they are part of and is the way that they can find out who they are working with
    * For authorized users (TAs and above), one or two buttons are available in the top right corner: edit and delete team. Only instructors can delete teams. Deleting a team will also delete its participation in the exercise (the text in the confirmation modal mentions this). When a team is deleted, the user is brought back to the team overview page.
6. New participation status `NO_TEAM_ASSIGNED`
    * When a team exercise has been released but the student has not yet been assigned to a team, he is not allowed to start the exercise. He is shown a message that he has to be assigned to a team first.
7. New button "Team" in student actions for exercise row
    * Once the student has been assigned to a team, he can start the exercise as usual. Another button is available for team-based exercises that will open up the Team detail page where all students of the team are listed (with their email for contacting them).
8. Nothing has changed
9. Nothing has changed
10. Participations table can now handle team-based exercises
    * For team based exercises, the participations show the team's name and students
    * Search also allows searching for both team name or student name/login
11. Nothing has changed
12. Exercise scores table can now handle team-based exercises
    * Shows team name and short name (name is linked to submissions as it is done for normal individual-mode exercises, short name is linked to Team detail page)

### Screenshots
1. Management of teams for an exercise
![OnPaste 20200315-184719](https://user-images.githubusercontent.com/7109225/76709581-1d73a300-6700-11ea-8c4e-ee2b8f6e6843.png)

2. Management of teams for an exercise: Improved search
![OnPaste 20200315-184757](https://user-images.githubusercontent.com/7109225/76709582-1e0c3980-6700-11ea-86e6-13108f9b618c.png)

3. Edit team dialog
![OnPaste 20200315-184838](https://user-images.githubusercontent.com/7109225/76709583-1ea4d000-6700-11ea-994e-c87c2186eed4.png)

4. Delete team confirmation modal
![OnPaste 20200315-184903](https://user-images.githubusercontent.com/7109225/76709584-1ea4d000-6700-11ea-914f-9b9c0f065d83.png)

5. Team detail page
![OnPaste 20200315-184943](https://user-images.githubusercontent.com/7109225/76709585-1f3d6680-6700-11ea-9dd4-2744edaba2df.png)

6. Participation status when student is not part of a team yet
![OnPaste 20200315-185027](https://user-images.githubusercontent.com/7109225/76709586-1f3d6680-6700-11ea-9bfe-b36158c047e3.png)

7. Team button available once assigned
![OnPaste 20200315-185127](https://user-images.githubusercontent.com/7109225/76709587-1fd5fd00-6700-11ea-8e60-c7dc07386b7e.png)

8. Team exercise has been started
![OnPaste 20200315-185226](https://user-images.githubusercontent.com/7109225/76709588-206e9380-6700-11ea-9cfa-f5e5179493b6.png)

9. Submission has been made for team exercise
![OnPaste 20200315-190451](https://user-images.githubusercontent.com/7109225/76709589-206e9380-6700-11ea-820d-0411fdab36bb.png)

10. Participations table for a team-based exercise
![OnPaste 20200315-191319](https://user-images.githubusercontent.com/7109225/76709591-206e9380-6700-11ea-8741-a3ba03243d7c.png)

11. Submissions for a team participation
![OnPaste 20200315-191439](https://user-images.githubusercontent.com/7109225/76709592-21072a00-6700-11ea-8fe9-5d1450b20877.png)

12. Exercise scores for a team exercise
![OnPaste 20200315-212512](https://user-images.githubusercontent.com/7109225/76709957-7bee5080-6703-11ea-9293-6f0d4bcf4406.png)

sleiss commented on src/main/webapp/i18n/de/exercise.json:
```diff
@@ -10,12 +10,12 @@
             "showAllResults": "Zeige alle Ergebnisse",
             "searchForStudents": "Suche nach Studenten Logins oder Namen (Komma getrennt)",
-            "searchForTeams": "Suche nach Team Namen (Komma getrennt)",
+            "searchForTeams": "Suche nach Team (Kurz-)Namen oder nach Studenten Logins oder Namen (Komma getrennt)",
```
Maybe we find a different translation here (maybe without two ***oder***), but this is really minor.

madwau replied:
Yeah, the two `oder` do make it sound a little more complicated than it probably has to be. I'll change it to `Suche nach Team (Kurz-)Namen oder nach Studenten Logins/Namen (Komma getrennt)`.

sleiss commented on src/main/webapp/i18n/de/exercise.json:
```diff
@@ -69,6 +69,8 @@
             "studentName": "Name",
             "studentId": "Login",
+            "teamName": "Team Name",
+            "teamShortName": "Kurzname",
```
Is it important to state that it's the short name of the team?

madwau replied:
You are right that this is a little inconsistent when looking at the translation of `teamName` vs `teamShortName`. The reason why I left out the `Team` in the translation of `teamShortName` is that it gets used right in the column header next to `Team Name`, so that it is obvious from context that the short name refers to the team and I wanted to save horizontal space since the column otherwise doesn't necessarily need that much width.

But it does feel like a bad practice to "cheat" with the translation for a specific use case. What I'll do: I'll rename this translation to `Team Kurzname` and add another translation pair just for `shortName` (if it doesn't already exist elsewhere). I think that's a good solution.

madwau replied:
Hm, on second thought that doesn't make too much sense either since the two student-related translations above also don't follow that kind of schema. I'd now probably tend to keep it as is where `teamName` is kind of the exception in that it contains the entity name in the translation value but I think that makes sense to differentiate it from the value of `studentName` (although in context it will probablt be obvious that those names are not student names but team names; the exception is when the table does not contain any rows).

sleiss commented on src/main/webapp/i18n/en/exercise.json:
```diff
@@ -10,7 +10,7 @@
             "showAllResults": "Show all results",
             "searchForStudents": "Search for students by login or name (comma separated)",
-            "searchForTeams": "Search for teams by name (comma separated)",
+            "searchForTeams": "Search for teams by (short) name or for students by login or name (comma separated)",
```
Same as in German.

madwau replied:
See comment above.

sleiss commented on src/main/webapp/i18n/en/exercise.json:
```diff
@@ -69,6 +69,8 @@
             "studentName": "Name",
             "studentId": "Login",
+            "teamName": "Team Name",
+            "teamShortName": "Short Name",
```
Same as in German.

madwau replied:
See comment above.

madwau commented on src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java:
```diff
@@ -648,37 +652,80 @@ public StudentParticipation findOneWithEagerSubmissionsAndResults(Long participa
-    public Optional<StudentParticipation> findOneByExerciseIdAndStudentLoginWithEagerSubmissionsAnyState(Long exerciseId, String username) {
-        log.debug("Request to get Participation for User {} for Exercise with id: {}", username, exerciseId);
-        return studentParticipationRepository.findWithEagerSubmissionsByExerciseIdAndStudentLogin(exerciseId, username);
+    public Optional<StudentParticipation> findOneByExerciseIdAndStudentLoginWithEagerSubmissionsAnyState(Exercise exercise, String username) {
```
Good catch, oversight on my part. I fixed it.

madwau commented on src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java:
```diff
@@ -735,12 +792,17 @@ public StudentParticipation findOneWithEagerSubmissionsAndResults(Long participa
      */
-    public List<StudentParticipation> findByExerciseIdAndStudentIdWithEagerResultsAndSubmissions(Long exerciseId, Long studentId) {
-        return studentParticipationRepository.findByExerciseIdAndStudentIdWithEagerResultsAndSubmissions(exerciseId, studentId);
+    public List<StudentParticipation> findByExerciseIdAndStudentIdWithEagerResultsAndSubmissions(Exercise exercise, Long studentId) {
```
Fixed.

madwau commented on src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java:
```diff
@@ -187,8 +187,7 @@ public ModelingSubmission lockModelingSubmissionWithoutResult(ModelingExercise m
     public ModelingSubmission save(ModelingSubmission modelingSubmission, ModelingExercise modelingExercise, String username) {
-        Optional<StudentParticipation> optionalParticipation = participationService.findOneByExerciseIdAndStudentLoginWithEagerSubmissionsAnyState(modelingExercise.getId(),
-                username);
+        Optional<StudentParticipation> optionalParticipation = participationService.findOneByExerciseIdAndStudentLoginWithEagerSubmissionsAnyState(modelingExercise, username);
```
See above (fixed).

madwau commented on src/main/java/de/tum/in/www1/artemis/service/FileUploadSubmissionService.java:
```diff
@@ -70,7 +70,7 @@ public FileUploadSubmissionService(FileUploadSubmissionRepository fileUploadSubm
     public FileUploadSubmission handleFileUploadSubmission(FileUploadSubmission fileUploadSubmission, MultipartFile file, FileUploadExercise fileUploadExercise,
             Principal principal) throws IOException, EmptyFileException {
-        Optional<StudentParticipation> optionalParticipation = participationService.findOneByExerciseIdAndStudentLoginAnyState(fileUploadExercise.getId(), principal.getName());
+        Optional<StudentParticipation> optionalParticipation = participationService.findOneByExerciseIdAndStudentLoginAnyState(fileUploadExercise, principal.getName());
```
See above (fixed).

madwau commented on src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:
```diff
@@ -430,7 +438,7 @@ public ResponseEntity getParticipationBuildArtifact(@PathVariable Long participa
         }
         else {
-            Optional<StudentParticipation> optionalParticipation = participationService.findOneByExerciseIdAndStudentLoginAnyStateWithEagerResults(exerciseId, principal.getName());
+            Optional<StudentParticipation> optionalParticipation = participationService.findOneByExerciseIdAndStudentLoginAnyStateWithEagerResults(exercise, principal.getName());
```
Fixed.

madwau commented on src/main/java/de/tum/in/www1/artemis/service/ComplaintService.java:
```diff
@@ -53,7 +53,7 @@ public Complaint createComplaint(Complaint complaint, Principal principal) {
                 .orElseThrow(() -> new BadRequestAlertException("The result you are referring to does not exist", ENTITY_NAME, "resultnotfound"));
         StudentParticipation studentParticipation = (StudentParticipation) originalResult.getParticipation();
-        User student = studentParticipation.getStudent();
+        User student = studentParticipation.getStudent().get(); // TODO: adapt for teams
```
Good question. I think adapting the complaints to teams is probably better suited to a follow-up PR since it might require some non-trivial decisions on how we want to handle complaints for team-based exercises. Also this PR is already on the larger side and since the creation of team-based exercises is still hidden for non-admin users, it is not like regular users will run into this under-construction part of the epic. I should still see about though that I deliver this within some days.

sleiss replied:
Alright, then let's handle this in a different PR.

sleiss commented:
I've tested the feature on the testserver with Text, Modeling and Programming exercises and it seems to work fine now (@madwau already fixed a small issue with the detail page)

madwau commented:
Thanks a lot for the feedback and for testing, @sleiss and @sascha11110!

madwau commented:
The test coverage for `TeamResource` is now at 100% (from 80% before). The overall test coverage for the REST controllers was increased from 89% on `develop` to 90% on this branch.

Non-trivial tests cases that were added:
- When searching for students to add to a team for an exercise, check that the boolean property `assignedToTeam` is true for those students who are already part of a team (this is important since otherwise already assigned students could be selected in the autocomplete search field and the user would see conflict information next to those students when he clicked on "create team"). The reverse is also tested that the property `assignedToTeam` is false for unassigned students.
- When requesting the courses with exercises for the dashboard and when requesting an exercise with details, make sure that the team exercise(s) contain(s) the transient attribute `studentAssignedTeamId` when the student has been assigned to a team for that exercise and that this id is correct (this is important for the participation status in the frontend; if it is not set, it is assumed that the student was not yet assigned to a team and they are shown the respective status message; only once that becomes true, can they start an exercise).
- Added a test to `ParticipationIntegrationTest`: Test what happens when a student tries to participate in an exercise when they have not yet been assigned to a team. The test succeeds if this correctly results in a bad request.

krusche commented:
> The test coverage for `TeamResource` is now at 100% (from 80% before). The overall test coverage for the REST controllers was increased from 89% on `develop` to 90% on this branch.
> 
> Non-trivial tests cases that were added:
> 
> * When searching for students to add to a team for an exercise, check that the boolean property `assignedToTeam` is true for those students who are already part of a team (this is important since otherwise already assigned students could be selected in the autocomplete search field and the user would see conflict information next to those students when he clicked on "create team"). The reverse is also tested that the property `assignedToTeam` is false for unassigned students.
> * When requesting the courses with exercises for the dashboard and when requesting an exercise with details, make sure that the team exercise(s) contain(s) the transient attribute `studentAssignedTeamId` when the student has been assigned to a team for that exercise and that this id is correct (this is important for the participation status in the frontend; if it is not set, it is assumed that the student was not yet assigned to a team and they are shown the respective status message; only once that becomes true, can they start an exercise).
> * Added a test to `ParticipationIntegrationTest`: Test what happens when a student tries to participate in an exercise when they have not yet been assigned to a team. The test succeeds if this correctly results in a bad request.

Sounds good. Can you also look into client side tests and increase the test coverage here as well?

madwau commented:
> Sounds good. Can you also look into client side tests and increase the test coverage here as well?

Yes, working on client-side tests now.

madwau commented on src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:
```diff
@@ -469,7 +469,7 @@ public void combineAllStudentCommits(Repository repository, ProgrammingExercise
             studentGit.reset().setMode(ResetCommand.ResetType.SOFT).setRef(latestHash.getName()).call();
             studentGit.add().addFilepattern(".").call();
-            var student = ((StudentParticipation) repository.getParticipation()).getStudent();
+            var student = ((StudentParticipation) repository.getParticipation()).getStudents().iterator().next();
```
Great catch. In theory it would indeed be possible that someone removed all students from a team that already has programming exercise participations before this method is called. Then this would have thrown a `NoSuchElementException`. I fixed it so that it uses the existing fallback correctly for empty teams.

madwau commented on src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseExportService.java:
```diff
@@ -87,8 +87,9 @@ public File exportStudentRepositories(long programmingExerciseId, @NotNull List<
             log.info("Request to export all student repositories of programming exercise " + programmingExerciseId + " with title '" + programmingExercise.getTitle() + "'");
         }
         else {
+            // TODO: Is this the expected behavior for teams? Or should all team members be listed instead?
```
The TODO placement might have been a little misleading; it refers to the whole method and not just the log statement. But I think you are right that only @krusche has enough insight here to know how `exportStudentRepositories` should behave for teams. The behavior right now is that it treats teams as it would treat students (and uses the team shortName for the student login).

krusche replied:
should be ok

madwau commented on src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseParticipationService.java:
```diff
@@ -104,17 +106,24 @@ public TemplateProgrammingExerciseParticipation findTemplateParticipationByProgr
     @NotNull
-    public ProgrammingExerciseStudentParticipation findStudentParticipationByExerciseIdAndStudentId(Long exerciseId, String username) throws EntityNotFoundException {
-        Optional<ProgrammingExerciseStudentParticipation> participation = studentParticipationRepository.findByExerciseIdAndStudentLogin(exerciseId, username);
+    public ProgrammingExerciseStudentParticipation findStudentParticipationByExerciseIdAndStudentId(Exercise exercise, String username) throws EntityNotFoundException {
```
Fixed

madwau commented on src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java:
```diff
@@ -123,9 +126,15 @@ else if (authCheckService.isStudentInCourse(course, user)) {
+                // for team-based exercises: add the team that the user has been assigned to onto the exercise
+                if (exercise.isTeamMode()) {
+                    teamService.findOneByExerciseAndUser(exercise, user).ifPresent(team -> {
+                        exercise.setStudentAssignedTeamId(team.getId());
```
I responded to that comment.

madwau commented:
Client-side tests that were added as part of this PR:
![OnPaste](https://user-images.githubusercontent.com/7109225/77335254-32ff5300-6d26-11ea-9c32-7f1005fb9cfd.png)

@krusche I also wanted to address the couple of little hiccups related to teams that we saw on `develop` during today's Skype call:
1. No feedback when clicking on "Delete" button for team (resulted in multiple clicks and alerts) => This has become obsolete with this PR since clicking the "Delete" button will now open our delete dialog component for confirmation.
2. When searching for your login, the student search did not return any result, which was probably correct since the user was likely not in the group of students, however, there was no indication apart from the very fast loading spinner that the search was actually performed and just didn't return any results. This is communicated more clearly now:
![OnPaste 20200323-165259](https://user-images.githubusercontent.com/7109225/77336854-6a6eff00-6d28-11ea-9a03-32bb5120b03b.png)
I also added a hint that the search terms needs to be at least three characters long (that appears when the user entered a search term that is too short) so that users aren't confused why the search doesn't trigger in that case:
![OnPaste 20200323-165215](https://user-images.githubusercontent.com/7109225/77337060-b02bc780-6d28-11ea-9482-89988b4d1834.png)
This is how it looks when the request failed (e.g. when the user has no internet connection):
![OnPaste 20200323-170823](https://user-images.githubusercontent.com/7109225/77337273-f8e38080-6d28-11ea-9883-8bb92916fad6.png)
3. The bug, that a student can't be removed from the pending team and then be added back right away, was already fixed as part of this PR. Screencast as proof:
![captured](https://user-images.githubusercontent.com/7109225/77337488-3fd17600-6d29-11ea-8e8d-4cbff64cbd01.gif)

sleiss commented on src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseBitbucketBambooIntegrationTest.java:
```diff
@@ -113,37 +132,44 @@ public void setup() throws Exception {
         testRepo.configureRepos("testLocalRepo", "testOriginRepo");
         solutionRepo.configureRepos("solutionLocalRepo", "solutionOriginRepo");
         studentRepo.configureRepos("studentRepo", "studentOriginaRepo");
+        studentTeamRepo.configureRepos("studentTeamRepo", "studentTeamOriginaRepo");
```
Is there a typo in `studentTeamOriginaRepo`?

madwau replied:
Definitely looks like a typo to me in the original `studentOriginaRepo` in the line above. I fixed both occurences.

sleiss replied:
Perfect!

madwau commented:
> I've tested the functionality, seems to work!
> Good job :)

Thanks! Also for reviewing and testing so thoroughly!

madwau commented:
> I tested it for a modeling and a programming exercise. Works great!
> 
> By the way: I like the UI, very intuitive! Although the feature is quite complex, I knew directly what I had to do.

Great to hear, that means a lot to me! Thanks again for testing :)

krusche commented on src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseExportService.java:
```diff
@@ -277,15 +278,15 @@ private void createZipFile(Path zipFilePath, List<Path> paths) throws IOExceptio
      */
     public void addStudentIdToProjectName(Repository repo, ProgrammingExercise programmingExercise, StudentParticipation participation) {
-        String studentId = participation.getStudent().getLogin();
+        String studentId = participation.getParticipantIdentifier();
```
I guess the variable `studentId`, should then be renamed to `participantId`

madwau replied:
Renamed it to `participantIdentifier`.

krusche commented on src/main/java/de/tum/in/www1/artemis/service/scheduled/AutomaticBuildPlanCleanupService.java:
```diff
@@ -67,7 +67,7 @@ public void cleanupBuildPlans() {
             }
-            if (participation.getStudent() == null) {
+            if (participation.getParticipant() == null) {
                 // we only want to clean up build plans of students
```
I guess we should update this comment

krusche replied:
```suggestion
                // we only want to clean up build plans of students or teams (NOT template or solution build plans)
```

madwau replied:
Updated

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:
```diff
@@ -567,8 +567,8 @@ public ProgrammingExerciseResource(ProgrammingExerciseRepository programmingExer
-            if (repositoryExportOptions.isExportAllStudents() || (programmingStudentParticipation.getRepositoryUrl() != null && studentParticipation.getStudent() != null
-                    && studentList.contains(studentParticipation.getStudent().getLogin()))) {
+            if (repositoryExportOptions.isExportAllStudents() || (programmingStudentParticipation.getRepositoryUrl() != null && studentParticipation.getParticipant() != null
+                    && studentParticipation.getStudents().stream().map(User::getLogin).anyMatch(studentList::contains))) {
```
I think there is a small logic error here. In case of a team exercise, we would like to export the repository only, if the team was mentioned in the "studentList", otherwise we would export the same repository multiple times, if the students of the same team are mentioned here.

So I would suggest we rename studentList to participantList and instructors need to refer to team short names for a team exercise and to student logins for individual exercises.

If you don't want to change this here, please add a TODO and fix it in a subsequent PR

madwau replied:
You are correct. Right now the same repository would get exported multiple times if more than one student from the same team is entered.

Your suggestion for how to handle this makes total sense. I'll include it here since it should be quick to adapt.

madwau replied:
Export for individual-mode exercises:
![OnPaste 20200329-153729](https://user-images.githubusercontent.com/7109225/77850598-eb6c4180-71d3-11ea-87c1-91d32c330507.png)

Export for team-mode exercises:
![OnPaste 20200329-153907](https://user-images.githubusercontent.com/7109225/77850600-ec04d800-71d3-11ea-8d5c-6937163815ee.png)

I tested the export for individual and team exercises with a specified list of participants and without (i.e. download all).

krusche commented on src/main/webapp/app/course/course-scores/course-scores.component.ts:
```diff
@@ -144,14 +144,17 @@ export class CourseScoresComponent implements OnInit, OnDestroy {
-            if (participation.presentationScore) {
-                student.presentationScore += participation.presentationScore;
+            const students = participation.student ? [participation.student] : participation.team.students;
+            for (let i = 0; i < students.length; i++) {
```
can we use a for each loop like here, something like the following?
(const student of students)

By the way, we should always prefer for each loops if they can be used

krusche replied:
One more question: are you sure this is enough code so that really all students in the team get the points for their final score? Also in a mixed scenario with some individual and some team participations?

madwau replied:
I switched the loop to a `for (const x of xs)` construct.

Yes, I just verified that this works as expected. In this scenario, the three marked students belong to the same team for a team-based text exercise. They received a result of 30/100 points for their submission. It was correctly credited to all of them and all of them were considered to have participated in this exercise.

![OnPaste 20200329-134143](https://user-images.githubusercontent.com/7109225/77848161-303bac80-71c3-11ea-9d4c-13dfe1aa6794.png)

krusche commented on src/main/webapp/app/overview/exercise-details/exercise-details-student-actions.component.html:
```diff
@@ -87,6 +87,15 @@
+            *ngIf="exercise.teamMode && participationStatusWrapper() !== ParticipationStatus.NO_TEAM_ASSIGNED"
+            [smallButton]="smallButtons"
+            [hideLabelMobile]="true"
+            [routerLink]="['/course-management', courseId, 'exercises', exercise.id, 'teams', assignedTeamId]"
```
Normal users should not be able to navigate to `course-management` URLs.
Can we offer the same page also on a different based on `courses/:courseId/exercises/...`

madwau replied:
Good point. I added a student-specific route at `courses/:courseId/exercises/:exerciseId/teams/:teamId` and restricted the other one to TAs and above.

krusche commented on src/main/webapp/app/overview/exercise-details/exercise-details-student-actions.component.html:
```diff
@@ -155,6 +164,16 @@
+            *ngIf="exercise.teamMode && participationStatusWrapper() !== ParticipationStatus.NO_TEAM_ASSIGNED"
+            [smallButton]="smallButtons"
+            [hideLabelMobile]="true"
+            [routerLink]="['/course-management', courseId, 'exercises', exercise.id, 'teams', assignedTeamId]"
```
Normal users should not be able to navigate to `course-management` URLs.
Can we offer the same page also on a different based on `courses/:courseId/exercises/...`

madwau replied:
See comment above.

krusche commented on src/main/webapp/app/overview/exercise-details/exercise-details-student-actions.component.html:
```diff
@@ -250,6 +269,15 @@
+            *ngIf="exercise.teamMode && participationStatusWrapper() !== ParticipationStatus.NO_TEAM_ASSIGNED"
+            [smallButton]="smallButtons"
+            [hideLabelMobile]="true"
+            [routerLink]="['/course-management', courseId, 'exercises', exercise.id, 'teams', assignedTeamId]"
```
Normal users should not be able to navigate to `course-management` URLs.
Can we offer the same page also on a different based on `courses/:courseId/exercises/...`

madwau replied:
See comment above.

krusche commented on src/main/webapp/app/overview/exercise-details/exercise-details-student-actions.component.html:
```diff
@@ -318,6 +346,15 @@
+            *ngIf="exercise.teamMode && participationStatusWrapper() !== ParticipationStatus.NO_TEAM_ASSIGNED"
+            [smallButton]="smallButtons"
+            [hideLabelMobile]="true"
+            [routerLink]="['/course-management', courseId, 'exercises', exercise.id, 'teams', assignedTeamId]"
```
Normal users should not be able to navigate to `course-management` URLs.
Can we offer the same page also on a different based on `courses/:courseId/exercises/...`

madwau replied:
See comment above.

krusche commented on src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java:
```diff
@@ -497,7 +505,7 @@ private ProgrammingExerciseStudentParticipation copyBuildPlan(ProgrammingExercis
             final var projectKey = participation.getProgrammingExercise().getProjectKey();
             final var planName = BuildPlanType.TEMPLATE.getName();
-            final var username = participation.getStudent().getLogin();
+            final var username = participation.getParticipantIdentifier();
```
rename username to participantIdentifier

madwau replied:
Renamed

krusche commented on src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java:
```diff
@@ -935,6 +1001,22 @@ public void deleteAllByExerciseId(Long exerciseId, boolean deleteBuildPlan, bool
+     */
+    @Transactional
+    public void deleteAllByTeamId(Long teamId, boolean deleteBuildPlan, boolean deleteRepository) {
+        List<StudentParticipation> participationsToDelete = findByTeamId(teamId);
```
please add a log.info(..) statement here so that we could potentially track this in the log

madwau replied:
Added

madwau commented:
> Great work 
> 
> I found some smaller issues in the code. Would be great if you can address them or add TODOs and address them in a subsequent PR.

Thanks! I addressed all comments and included the needed changes in this PR.
